### PR TITLE
{bio}[foss/2021a] OpenMM v7.7.0 w/ Python 3.9.5

### PR DESCRIPTION
**Pull request by i-mtz**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by i-mtz** [comment]
Test report by @i-mtz
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
sgi65.cluster.bc2.ch - Linux CentOS Linux 7.9.2009, x86_64, AMD EPYC 7742 64-Core Processor, 4 x NVIDIA A100-SXM4-40GB, 460.32.03, Python 2.7.5
See https://gist.github.com/08146d5dc0a8de91598e68f179228c25 for a full test report.

**Comment by i-mtz** [comment]
Test report by @i-mtz
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
sgi65.cluster.bc2.ch - Linux CentOS Linux 7.9.2009, x86_64, AMD EPYC 7742 64-Core Processor, 4 x NVIDIA A100-SXM4-40GB, 460.32.03, Python 2.7.5
See https://gist.github.com/233062c71a4ca0e3e560f9766e400cf4 for a full test report.

File: easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    ('Python', '3.9.5'),
+    ('SciPy-bundle', '2021.05'),
+    ('CUDA', '11.3.1', '', True),
+    ('SWIG', '4.0.2'),
+]
+
+# Flags for CUDA-enabled build
+configopts = "-DOPENMM_BUILD_CUDA_LIB=ON"
+
+#pretestopts = "export OPENMM_CUDA_COMPILER=nvcc && CTEST_OUTPUT_ON_FAILURE=1"
```
Comment (casparvl) [inline easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb:44]: Pretty sure this is causing the CI failure, as the `#` is not followed by a space. If this line is no longer relevant, please just take it out (or, if you commented it just for testing purposes, please uncomment it).

**Review by casparvl — COMMENTED** [review]
Overall, the PR looks good to me. The only thing being that commented line: I see that at least OpenMM-7.5.0-fosscuda-2020b.eb _did_ use it. If you think it is no longer needed, can you give a short explaination as to why it is obsolete now?

File: easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb (hunk 1 above)
Comment (i-mtz) [reply easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb:44]: I updated the PR removing the old `pretestopts` line. Initially I tried with the original `pretestopts` but it failed 1 test: `TestCudaCompiler`, from the OpenMM repo it seems that if you define `OPENM_CUDA_COMPILER=nvcc` will make this particular test to fail https://github.com/openmm/openmm/issues/3373 .

**Comment by i-mtz** [comment]
> Overall, the PR looks good to me. The only thing being that commented line: I see that at least OpenMM-7.5.0-fosscuda-2020b.eb _did_ use it. If you think it is no longer needed, can you give a short explaination as to why it is obsolete now?

I updated the PR removing the old `pretestopts` line. Initially I tried with the original `pretestopts` but it failed 1 test: `TestCudaCompiler`, from the OpenMM repo it seems that if you define `OPENM_CUDA_COMPILER=nvcc` will make this particular test to fail https://github.com/openmm/openmm/issues/3373 .

File: easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+# Author: Pablo Escobar Lopez
+# sciCORE - University of Basel
+# SIB Swiss Institute of Bioinformatics
+# Update to 7.5.1
+# J. Sassmannshausen / GSTT
+
+easyblock = 'CMakeMake'
+
+name = 'OpenMM'
+version = '7.7.0'
```
Comment (casparvl) [inline easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb:11]: I'm checking with other maintainers to be sure, but now that we don't use the fosscuda toolchain anymore, I think we always put a `versionsuffix = '-CUDA-%(cudaver)s'`. But as said, I'm still checking, I'll let you know once I'm sure.

**Comment by casparvl** [comment]
Test report by @casparvl
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
software1.lisa.surfsara.nl - Linux debian 10.12, x86_64, Intel(R) Xeon(R) Silver 4110 CPU @ 2.10GHz, Python 2.7.16
See https://gist.github.com/90df6e398341448d82dfdcc3d306025a for a full test report.

**Comment by casparvl** [comment]
Tests on `software1` are failing because that's a node without GPUs. Usually tests suites skip tests that are not supported by the current hardware, but I guess that's not true for OpenMM. You don't know if there's a simple way to skip the GPU based tests, right? If so we could consider if we should detect if the build node has GPUs, and otherwise run only the CPU based tests. Anyway, in my personal opinion, this should not be a blocker, but I'll check what other maintainers feel about that.

**Comment by casparvl** [comment]
Test report by @casparvl
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
gcn1 - Linux RHEL 8.4, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz, 4 x NVIDIA NVIDIA A100-SXM4-40GB, 470.57.02, Python 3.6.8
See https://gist.github.com/b6a691faa8d727125d95b16d5275b639 for a full test report.

File: easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb (hunk 2 above)
Comment (i-mtz) [reply easybuild/easyconfigs/o/OpenMM/OpenMM-7.7.0-foss-2021a.eb:11]: You are right, once I put this eb in production I realized I forgot about `versionsuffix = '-CUDA-%(cudaver)s'!

**Comment by i-mtz** [comment]
> Tests on `software1` are failing because that's a node without GPUs. Usually tests suites skip tests that are not supported by the current hardware, but I guess that's not true for OpenMM. You don't know if there's a simple way to skip the GPU based tests, right? If so we could consider if we should detect if the build node has GPUs, and otherwise run only the CPU based tests. Anyway, in my personal opinion, this should not be a blocker, but I'll check what other maintainers feel about that.

My understanding is that OpenMM will run checks for the GPU as we enable it on the config options: `configopts = "-DOPENMM_BUILD_CUDA_LIB=ON"`

There is also an stochastic [test](https://github.com/openmm/openmm/blob/be19e0222ddf66f612016a3c1f687161a53c2396/openmmapi/include/openmm/internal/AssertionUtilities.h#L59) that can randomly fail this causes the build to fail  which is not great for automated testing... I'm not sure about the best way to handle that one too...

**Comment by casparvl** [comment]
On `gcn1` I got:

```
99% tests passed, 1 tests failed out of 296

Total Test time (real) = 10814.04 sec

The following tests FAILED:
         67 - TestCudaCustomNonbondedForceSingle (Timeout)
Errors while running CTest
make: *** [Makefile:117: test] Error 8
```
the timeout suggests a hang...?

**Comment by i-mtz** [comment]
@casparvl by looking at the logs from the last run on `gcn1` with the A100 GPUs the time taken for each CUDA test seems significantly longer than in our case. In the builds I have tried the time for all tests takes around 3300s, the last test I did yesterday was with 64 cpus and 1 A100. Could it be some issue with the GPU?
```
58/296 Test  #58: TestCudaCustomGBForceSingle ...............   Passed   34.18 sec
        Start  59: TestCudaCustomGBForceMixed
 59/296 Test  #59: TestCudaCustomGBForceMixed ................   Passed   32.59 sec
        Start  60: TestCudaCustomGBForceDouble
 60/296 Test  #60: TestCudaCustomGBForceDouble ...............   Passed   36.68 sec
        Start  61: TestCudaCustomHbondForceSingle
 61/296 Test  #61: TestCudaCustomHbondForceSingle ............   Passed    5.70 sec
        Start  62: TestCudaCustomHbondForceMixed
 62/296 Test  #62: TestCudaCustomHbondForceMixed .............   Passed    6.36 sec
        Start  63: TestCudaCustomHbondForceDouble
 63/296 Test  #63: TestCudaCustomHbondForceDouble ............   Passed    6.72 sec
        Start  64: TestCudaCustomManyParticleForceSingle
 64/296 Test  #64: TestCudaCustomManyParticleForceSingle .....   Passed   28.73 sec
        Start  65: TestCudaCustomManyParticleForceMixed
 65/296 Test  #65: TestCudaCustomManyParticleForceMixed ......   Passed   28.14 sec
        Start  66: TestCudaCustomManyParticleForceDouble
 66/296 Test  #66: TestCudaCustomManyParticleForceDouble .....   Passed   30.27 sec
        Start  67: TestCudaCustomNonbondedForceSingle
 67/296 Test  #67: TestCudaCustomNonbondedForceSingle ........   Passed   50.67 sec
        Start  68: TestCudaCustomNonbondedForceMixed
 68/296 Test  #68: TestCudaCustomNonbondedForceMixed .........   Passed   49.91 sec
        Start  69: TestCudaCustomNonbondedForceDouble
 69/296 Test  #69: TestCudaCustomNonbondedForceDouble ........   Passed   52.28 sec
        Start  70: TestCudaCustomTorsionForceSingle
 70/296 Test  #70: TestCudaCustomTorsionForceSingle ..........   Passed    5.77 sec
        Start  71: TestCudaCustomTorsionForceMixed
 71/296 Test  #71: TestCudaCustomTorsionForceMixed ...........   Passed    6.26 sec
        Start  72: TestCudaCustomTorsionForceDouble
 72/296 Test  #72: TestCudaCustomTorsionForceDouble ..........   Passed    6.17 sec
        Start  73: TestCudaDispersionPMESingle
 73/296 Test  #73: TestCudaDispersionPMESingle ...............   Passed  121.99 sec
```

**Comment by casparvl** [comment]
Test report by @casparvl
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
software2.lisa.surfsara.nl - Linux debian 10.12, x86_64, Intel(R) Xeon(R) Bronze 3104 CPU @ 1.70GHz, 4 x NVIDIA NVIDIA TITAN V, 470.103.01, Python 2.7.16
See https://gist.github.com/0ad202eaddedadeb6ff3595dfd99c169 for a full test report.

**Comment by casparvl** [comment]
Test report by @casparvl
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
gcn1 - Linux RHEL 8.4, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz, 4 x NVIDIA NVIDIA A100-SXM4-40GB, 470.57.02, Python 3.6.8
See https://gist.github.com/f7e5cd904e1892e534b1b7fc04a0362f for a full test report.

**Comment by casparvl** [comment]
@i-mtz The reason the CI is still failing is that now the PR includes 2 EasyConfigs: your original one (without the `-CUDA` suffix and with the comment-issue that produces the CI failure) and one in which all of this is resolved. Unless you also want to make a CPU-only EasyConfig for OpenMM, I'm assuming you'll want to remove `OpenMM-7.7.0-foss-2021a.eb` from the PR. Please see [these instructions](https://docs.easybuild.io/en/latest/Integration_with_GitHub.html#deleting-easyconfig-files-or-patches) on how to delete files from your PR.

Regarding the tests: I reran it on `gcn1`, where this _one_ test failed earlier with a timeout, and also on one of our other GPU capable nodes. This time, everything passed, so I'm sure the first failure was just a hickup due to a slow filesystem on our side. Your comment that my runtimes where much longer than yours triggered me to check what was happening during the test suite. I saw a lot of small I/O (I think when it's compiling the test) and then the GPUs are only used for a _very_ short time. Thus, the I/O performance can have a big impact on the overall runtime - and apparently ours was pretty bad at the time I ran the test suite the first time.

Anyway, once the  `OpenMM-7.7.0-foss-2021a.eb` is removed, I think the CI will pass. I might then want to test the build one more time to make sure the latest commits didn't accidentally botch anything, but they shouldn't have. The PR should then be ready to be merged, almost there... :)

**Comment by i-mtz** [comment]
@casparvl Yes, I ended deleting/uploading the wrong EasyConfig, I reverted the change and now tests pass.

**Comment by casparvl** [comment]
Test report by @casparvl
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
gcn1 - Linux RHEL 8.4, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz, 4 x NVIDIA NVIDIA A100-SXM4-40GB, 470.57.02, Python 3.6.8
See https://gist.github.com/1c443e42f80631409316c6d918fc0a9e for a full test report.

**Review by casparvl — APPROVED** [review]
lgtm!

**Comment by casparvl** [comment]
Sorry @i-mtz , I lost sight of this one, even though we were practically at the end... All looks good now though, so I'll (finally) merge it. Thanks for the contribution!

**Pull request merged**